kubernetes/skaffold/src/main/kotlin/com/google/kubernetes/tools/skaffold/run/SkaffoldCommandLineState.kt

```diff
@@ -23,37 +23,95 @@ import com.google.kubernetes.tools.skaffold.message
- * window output and graceful process shutdown (also see [KillableProcessHandler])
+ * Implementation of [SkaffoldCommandLineState] for continuous deployment, "dev" mode.
+ */
+class SkaffoldDevCommandLineState(environment: ExecutionEnvironment) :
+        SkaffoldCommandLineState(environment) {
+
+    /**
+     * Returns the appropriate [SkaffoldExecutorSettings.ExecutionMode] base on the executor that is
+     * used for this environment. This depends on if the run configuration was started in "run" or
+     * "debug" mode.
+     */
+    override fun getExecutionMode():
+            SkaffoldExecutorSettings.ExecutionMode = when (environment.executor) {
+        is DefaultRunExecutor -> SkaffoldExecutorSettings.ExecutionMode.DEV
+        is DefaultDebugExecutor -> SkaffoldExecutorSettings.ExecutionMode.DEBUG
+        else -> throw RuntimeException("Unexpected Skaffold executor type found: " +
+                "${environment.executor}.")
+    }
+
+    /**
+     * Do not explicitly tail logs in "dev" mode since Skaffold will automatically stream the logs
+     * in the running process.
+     */
+    override fun shouldTailDeploymentLogs(runConfiguration: RunConfiguration): Boolean = false
+}
+
+/**
+ * Implementation of [SkaffoldCommandLineState] for single run deployment mode.
+ */
+class SkaffoldRunCommandLineState(environment: ExecutionEnvironment)
+    : SkaffoldCommandLineState(environment) {
+    private val log = Logger.getInstance(this::class.java)
+
+    /**
+     * Returns the single run [SkaffoldExecutorSettings.ExecutionMode].
+     */
+    override fun getExecutionMode(): SkaffoldExecutorSettings.ExecutionMode =
+            SkaffoldExecutorSettings.ExecutionMode.SINGLE_RUN
+
+    /**
+     * The deployment logs should be tailed if stored in the run configuration settings due to user
+     * selection.
+     */
+    override fun shouldTailDeploymentLogs(runConfiguration: RunConfiguration): Boolean =
+            if (runConfiguration is SkaffoldSingleRunConfiguration) {
+                runConfiguration.tailDeploymentLogs
+            } else {
+                log.warn("Unexpected Skaffold run configuration type detected for Skaffold run " +
+                        "mode. Will not tail logs")
+                false
+            }
+}
+
+/**
+ * Runs Skaffold on command line based on the given [AbstractSkaffoldRunConfiguration] and current
+ * IDE project. [startProcess] checks configuration and constructs command line to launch Skaffold
+ * process. Base class manages console window output and graceful process shutdown (also see
+ * [KillableProcessHandler])
  *
  * @param environment Execution environment provided by IDE
- * @param executionMode Skaffold  mode execution, i.e. DEV
  */
-class SkaffoldCommandLineState(
-    environment: ExecutionEnvironment,
-    val executionMode: SkaffoldExecutorSettings.ExecutionMode
+abstract class SkaffoldCommandLineState(
```

Comment:
I started looking through the new class structure and it seems for now the transformation is mostly for selecting either dev or debug mode - could this be done in `SkaffoldDevConfiguration.getState` instead - I think it has `environment` there so we can figure out run mode? Also, moving `tailLogs` as its own method seems to make it a bit harder to read - i.e., it's just part of the settings, and now it ends up being an own method for all command line states. I might miss something from follow-up PRs, we might need this structure for better debugger support?

Reply:
hmm. good point. I will refactor this. I'd rather introduce the changes as they are needed. Plus, on second look, I'm not convinced I'll need it anyway in the next PRs.

Reply:
done. ptal